production/sample/default/main.jsonnet

```diff
@@ -15,42 +14,122 @@ local tns_mixin = import 'tns-mixin/mixin.libsonnet';
     admin_services+: [
       { title: 'TNS Demo', path: 'tns-demo', url: 'http://app.tns.svc.cluster.local/', subfilter: true },
     ],
-    promtail_config+: {
-      clients: [{
-        username:: '',
-        password:: '',
-        scheme:: 'http',
-        hostname:: 'loki.loki.svc.cluster.local:3100',
-        external_labels: {
-          cluster: 'tns',
-        },
-      }],
-      pipeline_stages+: [
-        {
-          regex: {
-            expression: '\\((?P<status_code>\\d{3})\\)',
-          },
-        },
-        {
-          labels: {
-            status_code: '',
-          },
-        },
-        {
-          regex: {
-            expression: '(level|lvl|severity)=(?P<level>\\w+)',
+
+    // Allow the Grafana Agent to remote write directly to Prometheus. This allows us to use
+    // the Agent for all scraping duties and treat Prometheus as a remote write target similar
+    // to Cortex/GEM/Grafana Cloud.
+    prometheus_enabled_features+: [
+      'remote-write-receiver',
+    ],
+  },
+
+  local configMap = k.core.v1.configMap,
+  local container = k.core.v1.container,
+  local containerPort = k.core.v1.containerPort,
+  local envVar = k.core.v1.envVar,
+  local httpIngressPath = k.extensions.v1beta1.httpIngressPath,
+  local ingress = k.extensions.v1beta1.ingress,
+  local ingressRule = k.extensions.v1beta1.ingressRule,
+  local service = k.core.v1.service,
+
+  // Create a Grafana Agent daemon set to collect metrics, logs, and traces
+  //
+  // Metrics, logs, and traces are enriched with Kubernetes metadata. Metrics and logs from
+  // the local Kubernetes node are collected while traces use a push model (clients send traces
+  // to the agent).
+  //
+  // The agent runs as a privileged container and as root since this is a requirement of
+  // collecting logs. The path /var/log on the Kubernetes node is mounted into the container
+  // along with /var/lib/docker/containers.
+  //
+  // A service for the agent is created so that other pods within the cluster can send traces
+  // to the agent.
+  daemonset_agent:
+    gragent.new(name='grafana-agent', namespace='default') +
+    gragent.withDaemonSetController() +
+    gragent.withLogVolumeMounts({}) +
+    gragent.withLogPermissions({}) +
+    gragent.withConfigHash(true) +
+    gragent.withPortsMixin([
+      // Create container ports for the various ways that the agent can collect traces. Note that it's
+      // important to create the ports before creating the service for the agent since the method that
+      // creates the service will look at container ports to know which ports the service should have.
```

Comment:
Oh, really? That's surprising, I would've thought that given the merge logic in Jsonnet it doesn't matter which you declare first.

Reply:
I might be reading the Jsonnet wrong, but it seems like the `serviceFor` method here relies on the ports existing on the container already when creating the service: https://github.com/grafana/jsonnet-libs/blob/264a5c2078c5930af57fe2d107eff83ab63553af/ksonnet-util/util.libsonnet#L20

Reply:
That's true, but I thought it would only be evaluated post-merge and not from top-down.

Reply:
I'll try it out locally, changing up the order.

Reply:
You're right, I'll remove the comment.

Reply:
Fixed in e45543138433193c7ed21b5ff21807bd21c28c19